### PR TITLE
Use new `SpecificPartitionsImapReaderPms` for SQL IMap scan [HZ-2729]

### PR DESCRIPTION
**Pull request by Fly-Style**
Algorithm description here : https://github.com/hazelcast/hazelcast/pull/25031

### CODE REVIEW

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java [hunk 1]
```diff
@@ -232,14 +245,19 @@ public Vertex onFullScan(FullScanPhysicalRel rel) {
...
+        SqlConnector sqlConnector = getJetSqlConnector(table);
+        List<List<Expression<?>>> mapPartitionsToScanExprs = null;
+        if (sqlConnector.equals(IMapSqlConnector.INSTANCE)) {
+            mapPartitionsToScanExprs = computeRequiredPartitionsToScan(table.getSqlName());
+        }
+        return sqlConnector.fullScanReader(
                 dagBuildContext,
                 wrap(rel.filter()),
                 wrap(rel.projection()),
+                mapPartitionsToScanExprs,
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:257]: can't you pass `prunabilities.get(mapName);` and move rest of `computeRequiredPartitionsToScan` to IMap connector?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java [hunk 2]
```diff
@@ -48,20 +51,45 @@
...
         super.init(context);
-        Set<Integer> partitions = context.jobConfig().getArgument(KEY_REQUIRED_PARTITIONS);
-        if (partitions != null) {
-            partitionsToScan = partitions.stream().sorted().mapToInt(i -> i).toArray();
+        if (requiredPartitionsExprs != null) {
+            List<Integer> partitionsToScanList = new ArrayList<>();
+            partitionsToScan = new int[requiredPartitionsExprs.size()];
+            int j = 0;
+
+            HazelcastInstance hazelcastInstance = context.hazelcastInstance();
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java:73]: this logic repeats (similar may be possibly for index scan), in the future potential candidate for refactoring

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/CalciteSqlOptimizer.java [hunk 3]
```diff
@@ -890,8 +902,8 @@ private SqlRowMetadata createRowMetadata(
     private Tuple2<DAG, Set<PlanObjectKey>> createDag(
             PhysicalRel physicalRel,
             QueryParameterMetadata parameterMetadata,
-            Set<PlanObjectKey> usedViews
-    ) {
+            Set<PlanObjectKey> usedViews,
+            @Nullable Map<String, List<Map<String, Expression<?>>>> prunabilities) {
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/CalciteSqlOptimizer.java:906]: how many nested generic collections is a code smell?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java [hunk 4]
```diff
@@ -48,20 +51,45 @@
...
+                    partitionKeyComponents[i++] = expression.eval(null, eec);
+                }
+
+                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
+                        StrategyUtil.constructKey(partitionKeyComponents)
+                );
+                if (partition == null) {
+                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
+                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
+                    break;
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java:88]: `partitionsToScan` will be incomplete in this case instead of containing all patitions.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java (hunk 1 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:257]: Good catch btw, ty.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/CalciteSqlOptimizer.java (hunk 3 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/CalciteSqlOptimizer.java]: Not aware, sorry.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java (hunk 1 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java]: https://github.com/hazelcast/hazelcast/pull/25006/commits/71cfe45ba4334b96de3a4e540840d6efd81f937d

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java (hunk 4 above)
Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java:88]: this is not fixed

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            partitionsToScan = new int[requiredPartitionsExprs.size()];
+            int j = 0;
+
+            HazelcastInstance hazelcastInstance = context.hazelcastInstance();
+            ExpressionEvalContext eec = ExpressionEvalContext.from(context);
+            for (List<Expression<?>> requiredPartitionsExpr : requiredPartitionsExprs) {
+                Object[] partitionKeyComponents = new Object[requiredPartitionsExpr.size()];
+                int i = 0;
+                for (Expression<?> expression : requiredPartitionsExpr) {
+                    partitionKeyComponents[i++] = expression.eval(null, eec);
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java:79]: shouldn't we be using `evalTop`?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/QueryUtil.java [hunk 6]
```diff
@@ -49,7 +49,9 @@
 import java.util.List;
 import java.util.Map.Entry;
 
-final class QueryUtil {
+public final class QueryUtil {
+    public static final String DEFAULT_UNIQUE_KEY_TRANSFORMATION = "OBJECT";
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/QueryUtil.java:53]: this change does not belong to this PR. BTW, this class contains 2 completely unrelated groups of static methods used only in one place outside this class: one for joins, other one for index scan.

Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/QueryUtil.java:53]: It solves a code duplication issue. I moved these constants to the appropriate utility class. If you have a better placement idea - you may suggest them.

**Review by ivanthescientist — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                    partitionKeyComponents[i++] = expression.evalTop(null, eec);
+                }
+
+                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
+                        PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents)
+                );
+                if (partition == null) {
+                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
+                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
+                    return;
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java:88]: now in this case you fail to save `partitionAssignment` -> likely NPE in `get`

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/QueryUtil.java (hunk 6 above)
Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/QueryUtil.java:53]: `com.hazelcast.sql.impl.extract.KEY`.

`DEFAULT_UNIQUE_KEY_TRANSFORMATION` can stay where it were but also it might use `UniqueKeyTransformation.OBJECT.name`.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java [hunk 8]
```diff
@@ -236,14 +240,21 @@ public Vertex onFullScan(FullScanPhysicalRel rel) {
 
         dagBuildContext.setTable(table);
         dagBuildContext.setRel(rel);
-        return getJetSqlConnector(table).fullScanReader(
+
+        List<Map<String, Expression<?>>> partitionStrategyCandidate = null;
+        if (partitionStrategyCandidates != null) {
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:245]: scan pruning will not be used when member pruning is not possible. How you plan to fix that?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java (hunk 7 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java:88]: No, it will choose code path when `partitionsToScan` is null (actually, it's a default querying)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java (hunk 8 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:245]: Not sure if it is possible at all...

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java (hunk 7 above)
Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPms.java:88]: true, but a bit unexpectected. LocalProcessorSupplier.init will obtain partitionAssignment from the context on its own.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java (hunk 8 above)
Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:245]: this was a requirement. 

>Not sure if it is possible at all

It is: 6f4aa7cb464fc2b697180e9df2828b6f58c3eab2. This also demonstrates one of the limitations of current design, so it is not pretty, in wrong place and relies on some internals.

Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:245]: Ok, I got it : if you have only one prunable input for, let's say, UNION, and the second is not prunable -- member pruning is not applicable, but partition pruning is. Sorry, my bad.

Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:245]: this is one example, also interesting are joins: 8519c1c279df9f22c14868b4b497e35df5eabfc9

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/QueryUtil.java (hunk 6 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/QueryUtil.java]: Fixed https://github.com/hazelcast/hazelcast/pull/25006/commits/7d32b9ca07ee9c8a83d7d19158969dac085c5ab1

**Review by k-jamroz — APPROVED** [review]
